### PR TITLE
fix(datasets): per-instance feature-name mapping resolution

### PR DESCRIPTION
**Pull request by shuheng-liu**
## What this does

Fixes a silent feature-name-mapping collision between mixture entries. (🐛 Bug)

Two `dataset_mixture` entries that share both `repo_id` and `control_mode` but declare different `data_features_name_mapping` values — the natural pattern for training two camera views of the same repo, e.g. two `lerobot/droid_1.0.1` entries with `camera0 -> observation.images.exterior_1_left` vs `...exterior_2_left` — silently clobber each other today:

- `DatasetConfig.__post_init__` registers each mapping into the process-global `DATA_FEATURES_NAME_MAPPING` last-wins, and the composite `repo_id::control_mode` key only disambiguates entries with *different* control modes.
- At fetch time the datasets resolve their mapping from that registry, so **both instances read the last entry's columns** — the first entry's camera view never loads, and the mixture trains two identical copies of one view.

The fix stores the mapping per dataset instance and prefers it over the registry:

- `BaseDataset._get_name_map(strict=...)` resolves the per-instance mapping first, falling back to the registry (built-in defaults, annotation scripts, direct constructions keep working unchanged). All fetch-time lookups (`_to_standard_data_format`, `_load_subgoal_frames`, the `mistake`/`success` role resolution from #457) go through it.
- `resolve_delta_timestamps` prefers the entry's own mapping too, so each colliding entry's camera column gets its own temporal-window config.
- `DatasetMixtureMetadata` accepts per-entry `name_maps` (threaded from `WeightedDatasetMixture`, with the same one-level-deep unwrap as the `.meta` lookup for `random_split` wrappers), so normalization stats standardize against each entry's own columns; the `mistake`/`success` roles are excluded from the stats schema.
- `make_dataset` threads `cfg.data_features_name_mapping` into the `LeRobotDataset` constructor; the val-split clone shares it (shallow copy).
- Config registration still upserts the registry for its remaining consumers, but now **warns** when an entry overwrites the same *effective* key with a different mapping. The plain-`repo_id` fallback overwrite in the legitimate joint/ee dual-control-mode pattern stays silent — that one is by design.

No checked-in example config uses the colliding pattern (verified by scanning `configs/**/*.json`), so this only affects user configs.

## How it was tested

- Added `tests/datasets/test_name_mapping_collision.py`:
  - `_get_name_map`: instance mapping wins over the registry, registry fallback, strict `KeyError` vs non-strict empty-dict behavior.
  - The collision repro through the real constructor: two datasets, same `repo_id`, different `data_features_name_mapping` — each resolves its own `camera0`; the val-split clone shares the instance mapping.
  - Config-time warning: fires on a conflicting same-key registration (registry stays last-wins), silent for identical re-registration and for the joint/ee different-control-mode pattern.
  - `resolve_delta_timestamps`: each colliding entry's camera column gets its own deltas; `DatasetMixtureMetadata`: colliding entries standardize their own camera stats, and a mapped `success` column stays out of the stats schema; `make_dataset`: the constructor receives each entry's own mapping.
- `uv run pytest tests/datasets/ tests/configs/ -m "not gpu and not network" -n auto` → 722 passed, 7 skipped (includes the pre-existing `test_feature_mapping_dual_split.py` coverage of the joint/ee registry pattern).
- `pre-commit run` on all changed files → clean.

## How to checkout & try? (for the reviewer)

```bash
pytest -sx tests/datasets/test_name_mapping_collision.py
```

Or reproduce the pre-fix behavior on any config with two same-repo entries mapping different `camera0` columns: before this change both datasets return the same `_get_name_map()["camera0"]`; after it, each returns its own.

## Checklist

- [x] I have added Google-style docstrings to important functions and ensured function parameters are typed.
- [ ] My PR includes policy-related changes.
  - [ ] If the above is checked: I have run the GPU pytests (pytest -m "gpu") and regression tests.

### CODE REVIEW

File: src/opentau/datasets/factory.py [hunk 1]
```diff
@@ -273,6 +273,7 @@ def make_dataset(
             return_advantage_input=return_advantage_input,
             skip_timestamp_check=effective_skip,
             prompt_substitutions=cfg.prompt_substitutions,
+            data_features_name_mapping=cfg.data_features_name_mapping,
```
Comment (shuheng-liu) [inline src/opentau/datasets/factory.py:286]: Fixed in 4926ab2 — `resolve_delta_timestamps` now prefers `dataset_cfg.data_features_name_mapping` when set (registry only as fallback), with a collision test asserting each entry's camera column gets its own delta timestamps. Same commit also threads per-entry mappings into `DatasetMixtureMetadata` (via a `name_maps` param from `WeightedDatasetMixture`, one-level-deep unwrap for random_split wrappers) so normalization stats standardize against each entry's own columns, and excludes the `mistake`/`success` roles from the stats schema. Plus a `make_dataset` test asserting the constructor receives each colliding entry's own mapping.

File: tests/datasets/test_name_mapping_collision.py [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+        import warnings as _warnings
+
+        with _warnings.catch_warnings():
+            _warnings.simplefilter("error")
+            DatasetConfig(
+                repo_id="_tests/collision_repo",
+                control_mode="ee",
+                data_features_name_mapping={"actions": "action_ee"},
+            )
+        DATA_FEATURES_NAME_MAPPING.pop("_tests/collision_repo::ee", None)
```
Comment (shuheng-liu) [inline tests/datasets/test_name_mapping_collision.py]: Fixed in 4926ab2 — `_tests/collision_repo::ee` is now in `_TEST_KEYS` and cleaned by the autouse fixture; the manual pop is gone.

File: src/opentau/configs/default.py [hunk 3]
```diff
@@ -243,17 +249,37 @@ def __post_init__(self):
...
-                    DATA_FEATURES_NAME_MAPPING[key] = self.data_features_name_mapping
+                effective = feature_mapping_key(self.repo_id, self.control_mode)
+                keys = list(dict.fromkeys([self.repo_id, effective]))
+            for key in keys:
+                # Only the effective (composite-or-plain) key is what this
+                # entry's datasets resolve; overwriting the plain repo_id
+                # back-compat slot from a different control mode is by design.
+                previous = DATA_FEATURES_NAME_MAPPING.get(key)
+                if key == effective and previous is not None and previous != self.data_features_name_mapping:
+                    warnings.warn(
```
[the remark this answers is not in the document]
Comment (shuheng-liu) [reply src/opentau/configs/default.py:291]: Fixed in ff3f485 — registrations from config entries are now tracked in a module-level key set, and the warning fires only when the previous registration for the effective key came from another entry. Overriding a built-in default stays silent (covered by a new test alongside the existing conflict/identical/joint-ee cases).

File: src/opentau/configs/default.py [hunk 4]
```diff
@@ -243,17 +254,47 @@ def __post_init__(self):
...
+                # overriding a *built-in* default mapping, is by design and
+                # stays silent.
+                previous = DATA_FEATURES_NAME_MAPPING.get(key)
+                if (
+                    key == effective
+                    and key in _CONFIG_REGISTERED_MAPPING_KEYS
+                    and previous is not None
+                    and previous != self.data_features_name_mapping
+                ):
+                    warnings.warn(
```
[the remark this answers is not in the document]
Comment (shuheng-liu) [reply src/opentau/configs/default.py:291]: Fixed in 9d5eb1b — the set now tracks *effective* keys (what an entry actually resolves through), and the warning fires on any different-mapping overwrite of a tracked key from either the effective or the back-compat write. Both directions covered by tests: joint-then-None stays silent (only the back-compat slot is touched), None-then-joint now warns (the plain-key entry's effective registration is clobbered). The diagnostic suggestion is also addressed in the same commit: `diagnose_norm_distribution` prefers each entry's own mapping for column resolution and passes `name_maps` into `DatasetMixtureMetadata`.

File: src/opentau/datasets/lerobot_dataset.py [hunk 5]
```diff
@@ -760,6 +760,36 @@ class BaseDataset(torch.utils.data.Dataset):
...
     """
 
+    # Per-instance feature-name mapping. When set (LeRobotDataset receives it
+    # from the mixture entry's `data_features_name_mapping`), it wins over the
+    # process-global DATA_FEATURES_NAME_MAPPING registry — two mixture entries
+    # sharing a repo_id and control_mode would otherwise clobber each other's
+    # registry entry (last-wins) and silently read the same columns.
+    _data_features_name_mapping: dict[str, str] | None = None
+
+    def _get_name_map(self, strict: bool = True) -> dict[str, str]:
```
[the remark this answers is not in the document]
Comment (shuheng-liu) [reply src/opentau/datasets/lerobot_dataset.py:770]: Agreed this is real but follow-up scope: `source` is resolved inside `BaseDataset`, while the deduplicated per-entry name (`repo_id#N`) only exists at the `_TaggedDataset` wrapping layer — threading it down touches the tagging architecture rather than the mapping resolution this PR fixes, and per-sample `dataset_index` provenance already disambiguates the colliding entries everywhere except the outlier-debug merge key. Tracking it as a follow-up.
